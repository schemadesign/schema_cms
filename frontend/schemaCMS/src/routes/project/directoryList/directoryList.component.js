import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import Helmet from 'react-helmet';
import { FormattedMessage } from 'react-intl';
import { always, cond, path, propEq, T } from 'ramda';
import { Typography } from 'schemaUI';

import { Container } from './directoryList.styles';
import messages from './directoryList.messages';
import { TopHeader } from '../../../shared/components/topHeader';
import { ProjectTabs } from '../../../shared/components/projectTabs';
import { DIRECTORY } from '../../../shared/components/projectTabs/projectTabs.constants';
import { ContextHeader } from '../../../shared/components/contextHeader';
import { BackArrowButton, NavigationContainer, PlusButton } from '../../../shared/components/navigation';
import { Loader } from '../../../shared/components/loader';
import extendedDayjs, { BASE_DATE_FORMAT } from '../../../shared/utils/extendedDayjs';
import { ListContainer, ListItem } from '../../../shared/components/listComponents';
import { HeaderItem, HeaderList, titleStyles } from '../list/list.styles';

const { H1 } = Typography;

export class DirectoryList extends PureComponent {
  static propTypes = {
    directories: PropTypes.array.isRequired,
    fetchList: PropTypes.func.isRequired,
    intl: PropTypes.object.isRequired,
    match: PropTypes.shape({
      params: PropTypes.shape({
        projectId: PropTypes.string.isRequired,
      }),
    }),
    history: PropTypes.shape({
      push: PropTypes.func.isRequired,
    }),
  };

  state = {
    loading: true,
  };

  async componentDidMount() {
    try {
      const projectId = this.getProjectId();
      await this.props.fetchList({ projectId });
      this.setState({ loading: false });
    } catch (e) {
      this.props.history.push('/');
    }
  }

  getProjectId = () => path(['match', 'params', 'projectId'], this.props);

  handleCreateDirectory = () => this.props.history.push(`/project/${this.getProjectId()}/create`);

  handleShowProject = () => this.props.history.push(`/project/${this.getProjectId()}`);

  handleShowDirectory = id => this.props.history.push(`/directory/${id}`);

  renderHeader = (list = []) => (
    <HeaderList>
      {list.map((item, index) => (
        <HeaderItem id={`headerItem-${index}`} key={index}>
          {item}
        </HeaderItem>
      ))}
    </HeaderList>
  );

  renderItem({ id, name = '', created = '', createdBy = {} }, index) {
    const { firstName, lastName } = createdBy;
    const whenCreated = extendedDayjs(created, BASE_DATE_FORMAT).fromNow();
    const header = this.renderHeader([whenCreated, `${firstName}${lastName}`]);

    return (
      <ListItem key={index} headerComponent={header}>
        <H1 id={`projectName-${index}`} customStyles={titleStyles} onClick={() => this.handleShowDirectory(id)}>
          {name}
        </H1>
      </ListItem>
    );
  }

  renderList = ({ list }) => <ListContainer>{list.map((item, index) => this.renderItem(item, index))}</ListContainer>;

  renderContent = cond([
    [propEq('loading', true), always(<Loader />)],
    [propEq('list', []), this.renderNoData],
    [T, this.renderList],
  ]);

  render() {
    const { match, directories } = this.props;
    const { loading } = this.state;
    const headerTitle = <FormattedMessage {...messages.title} />;
    const headerSubtitle = <FormattedMessage {...messages.subTitle} />;

    return (
      <Container>
        <Helmet title={this.props.intl.formatMessage(messages.title)} />
        <TopHeader headerTitle={headerTitle} headerSubtitle={headerSubtitle} />
        <ProjectTabs active={DIRECTORY} url={`/project/${match.params.projectId}`} />
        <ContextHeader title={headerTitle} subtitle={headerSubtitle}>
          <PlusButton id="createDirectoryDesktopBtn" onClick={this.handleCreateDirectory} />
        </ContextHeader>
        {this.renderContent({ loading, list: directories })}
        <NavigationContainer hideOnDesktop>
          <BackArrowButton id="backBtn" onClick={this.handleShowProject} />
          <PlusButton id="createDirectoryBtn" onClick={this.handleCreateDataSource} />
        </NavigationContainer>
      </Container>
    );
  }
}
