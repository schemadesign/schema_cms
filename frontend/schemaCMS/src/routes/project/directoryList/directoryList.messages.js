/* eslint-disable max-len */
import { defineMessages } from 'react-intl';

export default defineMessages({
  title: {
    id: 'project.directoryList.title',
    defaultMessage: 'Directories',
  },
  subTitle: {
    id: 'project.directoryList.subTitle',
    defaultMessage: 'Overview',
  },
});
