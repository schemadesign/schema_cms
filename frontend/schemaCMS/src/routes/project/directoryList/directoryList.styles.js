import styled from 'styled-components';

export const Container = styled.div``;

export const ListItemContent = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  align-items: flex-start;
  height: 100%;
`;
