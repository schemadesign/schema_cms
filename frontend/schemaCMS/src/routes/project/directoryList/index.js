export { default as DirectoryList } from './directoryList.container';
