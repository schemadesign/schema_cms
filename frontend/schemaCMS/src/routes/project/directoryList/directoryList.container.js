import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { withRouter } from 'react-router-dom';
import { hot } from 'react-hot-loader';
import { compose } from 'ramda';
import { injectIntl } from 'react-intl';
import { bindPromiseCreators, promisifyRoutine } from 'redux-saga-routines';

import { DirectoryList } from './directoryList.component';
import { selectDirectories, DirectoryRoutines } from '../../../modules/directory';

const mapStateToProps = createStructuredSelector({
  directories: selectDirectories,
});

export const mapDispatchToProps = dispatch => ({
  ...bindPromiseCreators(
    {
      fetchList: promisifyRoutine(DirectoryRoutines.fetchList),
    },
    dispatch
  ),
});

export default compose(
  hot(module),
  connect(
    mapStateToProps,
    mapDispatchToProps
  ),
  injectIntl,
  withRouter
)(DirectoryList);
