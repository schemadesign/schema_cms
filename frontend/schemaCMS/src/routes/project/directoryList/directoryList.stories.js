import React from 'react';
import { storiesOf } from '@storybook/react';

import { DirectoryList } from './directoryList.component';

const defaultProps = {};

storiesOf('DirectoryList', module).add('Default', () => <DirectoryList {...defaultProps} />);
