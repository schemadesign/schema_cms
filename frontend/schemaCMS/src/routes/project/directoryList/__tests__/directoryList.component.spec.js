import React from 'react';
import { shallow } from 'enzyme';

import { DirectoryList } from '../directoryList.component';

describe('DirectoryList: Component', () => {
  const defaultProps = {};

  const component = props => <DirectoryList {...defaultProps} {...props} />;

  const render = (props = {}) => shallow(component(props));

  it('should render correctly', () => {
    const wrapper = render();
    global.expect(wrapper).toMatchSnapshot();
  });
});
