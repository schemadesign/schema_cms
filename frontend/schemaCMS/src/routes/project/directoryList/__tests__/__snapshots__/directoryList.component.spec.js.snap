// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DirectoryList: Component should render correctly 1`] = `
<styled.div>
  <HelmetWrapper
    defer={true}
    encodeSpecialCharacters={true}
    title="Directories"
  />
  <TopHeader
    headerSubtitle={
      <FormattedMessage
        defaultMessage="Overview"
        id="project.directoryList.subTitle"
        values={Object {}}
      />
    }
    headerTitle={
      <FormattedMessage
        defaultMessage="Directories"
        id="project.directoryList.title"
        values={Object {}}
      />
    }
  />
  <ProjectTabs
    active="directory"
    url="/project/1"
  />
  <ContextHeader
    subtitle={
      <FormattedMessage
        defaultMessage="Overview"
        id="project.directoryList.subTitle"
        values={Object {}}
      />
    }
    title={
      <FormattedMessage
        defaultMessage="Directories"
        id="project.directoryList.title"
        values={Object {}}
      />
    }
  >
    <PlusButton
      id="createDirectoryDesktopBtn"
      onClick={[Function]}
    />
  </ContextHeader>
  <LoadingWrapper
    error={null}
    loading={false}
    noData={false}
    noDataContent={null}
  >
    <styled.section>
      <Styled(Component)
        headerComponent={
          <ForwardRef(styled.ul)>
            <ForwardRef(styled.li)
              id="headerItem-0"
            >
              12 days ago
            </ForwardRef(styled.li)>
            <ForwardRef(styled.li)
              id="headerItem-1"
            >
              firstNamelastName
            </ForwardRef(styled.li)>
          </ForwardRef(styled.ul)>
        }
        key="0"
      >
        <H1
          customStyles={
            Object {
              "cursor": "pointer",
            }
          }
          id="projectName-0"
          onClick={[Function]}
        >
          name
        </H1>
      </Styled(Component)>
    </styled.section>
  </LoadingWrapper>
  <NavigationContainer
    hideOnDesktop={true}
    right={false}
  >
    <BackArrowButton
      id="backBtn"
      onClick={[Function]}
    />
    <PlusButton
      id="createDirectoryBtn"
    />
  </NavigationContainer>
</styled.div>
`;

exports[`DirectoryList: Component should render correctly with loader 1`] = `
<styled.div>
  <HelmetWrapper
    defer={true}
    encodeSpecialCharacters={true}
    title="Directories"
  />
  <TopHeader
    headerSubtitle={
      <FormattedMessage
        defaultMessage="Overview"
        id="project.directoryList.subTitle"
        values={Object {}}
      />
    }
    headerTitle={
      <FormattedMessage
        defaultMessage="Directories"
        id="project.directoryList.title"
        values={Object {}}
      />
    }
  />
  <ProjectTabs
    active="directory"
    url="/project/1"
  />
  <ContextHeader
    subtitle={
      <FormattedMessage
        defaultMessage="Overview"
        id="project.directoryList.subTitle"
        values={Object {}}
      />
    }
    title={
      <FormattedMessage
        defaultMessage="Directories"
        id="project.directoryList.title"
        values={Object {}}
      />
    }
  >
    <PlusButton
      id="createDirectoryDesktopBtn"
      onClick={[Function]}
    />
  </ContextHeader>
  <LoadingWrapper
    error={null}
    loading={true}
    noData={false}
    noDataContent={null}
  >
    <styled.section>
      <Styled(Component)
        headerComponent={
          <ForwardRef(styled.ul)>
            <ForwardRef(styled.li)
              id="headerItem-0"
            >
              12 days ago
            </ForwardRef(styled.li)>
            <ForwardRef(styled.li)
              id="headerItem-1"
            >
              firstNamelastName
            </ForwardRef(styled.li)>
          </ForwardRef(styled.ul)>
        }
        key="0"
      >
        <H1
          customStyles={
            Object {
              "cursor": "pointer",
            }
          }
          id="projectName-0"
          onClick={[Function]}
        >
          name
        </H1>
      </Styled(Component)>
    </styled.section>
  </LoadingWrapper>
  <NavigationContainer
    hideOnDesktop={true}
    right={false}
  >
    <BackArrowButton
      id="backBtn"
      onClick={[Function]}
    />
    <PlusButton
      id="createDirectoryBtn"
    />
  </NavigationContainer>
</styled.div>
`;
